improver/utilities/field_texture.py

```diff
@@ -0,0 +1,208 @@
+from improver.threshold import BasicThreshold
+from improver.utilities.cube_manipulation import collapsed
+
+
+class FieldTexture(BasePlugin):
+    """Plugin to calculate the texture in a binary field considering edge transitions
+
+    1) Takes a binary field that has been thresholded and looks for the transitions/edges
+       in the field that mark out a transition.
+    2) The transition calculation is then fed into the neighbourhooding code
+       (_calculate_ratio) to calculate a ratio for each cell.
+    3) The new cube of ratios is then thresholded and the realization coordinates
+       are collapsed to generate a mean of the thresholded ratios.
+    """
+
+    def __init__(self, nbhood_radius, ratio_threshold):
+        """
+
+        Args:
+            nbhood_radius (float):
+                The neighbourhood radius in metres within which the number of potential
+                transitions should be calculated. This forms the denominator in the
+                calculation of the ratio of actual to potential transitions that indicates a
+                field's texture. A larger radius should be used for diagnosing larger-scale
+                textural features.
+
+            ratio_threshold (float):
+                A unit-less threshold value that defines the ratio value above which
+                the field is considered clumpy and below which the field is considered
+                more contiguous.
+
+        """
+        self.nbhood_radius = nbhood_radius
+        self.ratio_threshold = ratio_threshold
+
+    def _calculate_ratio(self, cube, radius):
+        """
+        Calculates the ratio of actual to potential value transitions in a
+        neighbourhood about each cell.
+
+        The process is as follows:
+
+            1. For each grid cell find the number of cells of value 1 in a surrounding
+               neighbourhood of a size defined by the arg radius. The potential
+               transitions within that neighbourhood are defined as the number of
+               orthogonal neighbours (up, down, left, right) about cells of value 1.
+               This is 4 times the number of cells of value 1.
+            2. Calculate the number of actual transitions within the neighbourhood,
+               that is the number of cells of value 0 that orthogonally abut cells
+               of value 1.
+            3. Calculate the ratio of actual to potential transitions.
+
+        Ratios approaching 1 indicate that there are many transitions, so the field
+        is highly textured. Ratios close to 0 indicate a smoother field.
+
+        A neighbourhood full of cells of value 1 will return ratios of 0; the
+        diagnostic that has been thresholded to produce the binary field is found
+        everywhere within that neighbourhood, giving a smooth field. At the other
+        extreme, in neighbourhoods in which there are no cells of value 1 the ratio
+        is set to 1.
```

Comment:
Is there a simple explanation as to why the field is considered "smooth" if it exceeds the threshold everywhere, but "rough" if it doesn't?